docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,137 @@
+
+While in incremental data replication scenario the congestion does not happen often, but there still exists some use scenario that may lead to congestion, including
+
+- User pauses replication task for a while and a lot of relay log is accumulated, and the replication node has high concurrency.
+- In shard scenario and there are too many upstream shard instances replicating to downstream at the same time.
+- In none shard scenario, but there are many tasks replicating to downstream from multiple instances at the same time.
+
+When we encounter these abnormal scenario, we often pause part of the tasks or decrease the SQL executor concurrency. The key point of solving the congestion problem is to reduce the data flow speed. Currently DM has no data flow control, we need to build a data flow control framework which will make data import or replication more smooth, more automotive and as fast as possible. The data flow control framework should provide the following features:
+
+- Downstream congestion definition and auto detection
+- Import or replication speed auto control
+
+## Implementation
+
+### Congestion definition
+
+There exists some key concepts in a data import/replication link:
+
+- Data loss: Each database transaction execution failure is treated as one data loss, the failure includes every kind of database execution failure that no data is written to downstream, partial data written error does not include.
+- Bandwidth: Bandwidth describes the maximum data import/replication rate from a DM-worker to downstream, this can be measured by TPS from DM-worker (or TPS in downstream, but should only include the data traffic from this DM-worker).
+- Latency: Latency is the amount of time it takes for data to replicate from DM-worker to downstream.
+
+The congestion usually means the quality of service decreases because of the service node or data link is carrying more data than it can handle. In the data import/replication scenario, the consequence of congestion can be partial downstream database execution timeout, downstream qps decrease or SQL execution latency increase. We can use these three indicators to determine whether congestion happens and measure the degree of congestion.
+
+#### How to measure the key indicators
+
+Before we can measure the congestion, we should find a way to estimate the data loss, the bandwidth and the latency.
+
+- Data loss: can be collected in DM-worker itself, and we often use it as data loss rate, which means the count of sqls in all failed transaction divided by the count sqls in all transactions that need to be processed.
+- Latency: equals to the transaction executed latency, can be collected in DM-worker itself.
+- Bandwidth: this is a little complicated, but we can estimate it by adjusting the concurrency and find the maximum downstream TPS for this DM-worker, which can be used as the bandwidth of the data import/replication link.
```

Comment:
we can use the formula `qps = tps * batch = (1/avg lantency)*concurrency*(batch size)`